File: DataReconcile_Matlab/Uncertainty/FuzzyInterval.m
% Преобразование разных типов представления неточных данных к формату,
% унифицированному для дальнейшего применения Data Reconciliation.

% Класс для переменных типа FuzzyInterval (нечеткий интервал)
classdef FuzzyInterval
    properties
        AlphaLevels  % уровни α-cut (вектор 1×N)
        Intervals    % соответствующие вложенные интервалы (матрица N×2)
    end
    
    methods
        function obj = FuzzyInterval(alphaLevels, intervals)
            % Проверка входных аргументов
            if size(intervals,2) ~= 2
                error('Границы вложенных интервалов должны задаваться строкой из двух значений.');
            end
            if length(alphaLevels) ~= size(intervals, 1)
                error('Количество уровней alpha-cut должно соответствовать числу вложенных интервалов.');
            end
            if any(alphaLevels < 0 | alphaLevels > 1)
                error('Уровни alpha-cut должны иметь значения в пределах от 0 до 1.');
            end
            if any(intervals(:,1) > intervals(:,2))
                error('Правые границы вложенных интервалов должны быть меньше левых границ.');
            end
            
            % Сортировка по значения alpha-cut
            [obj.AlphaLevels, idx] = sort(alphaLevels(:)', 'descend');
            obj.Intervals = intervals(idx,:);

            % Проверка вложенного характера интервалов Intervals:
            % с уменьшением α-cut вложенные интервалы должны расширяться.
            if any(obj.Intervals(:,1)-sort(obj.Intervals(:,1),'descend')) || ...
               any(obj.Intervals(:,2)-sort(obj.Intervals(:,2),'ascend'))     
                error('Вложенность интервалов в поле Intervals нарушена.');
            end
        end
        
        % определение величины возможного среднеквадратического отклонения (согласно теореме Крейновича-Тао).
        function stdCint = getStd(obj, method)
            int = obj.Intervals;    % переменная для краткости
            a = obj.AlphaLevels;    % переменная для краткости
            widths = int(:,2)-int(:,1);     % ширина интервалов

            if (nargin < 2) || ((nargin == 2) && strcmp(method,'approximate'))
                alpha = 0.05;
                ind1 = find(a == alpha, 1, 'first');
                ind2 = find(a == 1, 1, 'first');
                if ~isempty(ind1) && ~isempty(ind2)
                   stdCint = [0, 0.25*(widths(ind1)-widths(ind2))];
                else
                   ind1 = find(a>alpha, 1, 'last');
                   if ind1 == 1
                      stdCint = [0, 0.25*widths(ind1)];
                   else
                      % линейная интерполяция
                      width = (alpha-a(ind1-1)) * (widths(ind1)-widths(ind1-1)) / (a(ind1)-a(ind1-1)) + widths(ind1-1);
                      stdCint = [0, 0.25*width];
                   end
                   if ~isempty(ind2)
                      stdCint(2) = stdCint(2) - 0.25*widths(ind2);
                   end
                end
            
            elseif strcmp(method,'accurate')
                % дискретизация носителя, используя границы вложенных интервалов
                breakpoints = unique([obj.Intervals(:,1); obj.Intervals(:,2)]);
                nBins = length(breakpoints) - 1;
                binEdges = sort(breakpoints);
                
                % определение ограничений на вероятности из α-cuts
                A = [];
                b = [];
                for i = 1:length(obj.AlphaLevels)
                    alpha = obj.AlphaLevels(i);
                    a = obj.Intervals(i,1);
                    b_int = obj.Intervals(i,2);
                    inBin = (binEdges(1:end-1) >= a) & (binEdges(2:end) <= b_int);
                    
                    % добавление ограничения: sum(p(inBin)) >= 1 - alpha
                    A = [A; -double(inBin')];  % Преобразование к виду sum(p)>=1-alpha
                    b = [b; -(1-alpha)];
                end
                
                % добавление ограничений на сумму вероятностей
                Aeq = ones(1, nBins);
                beq = 1;
                % добавление ограничения на неотрицательность вероятностей
                lb = zeros(nBins, 1);
                ub = [];
                
                % параметры промежутков
                binMids = (binEdges(1:end-1) + binEdges(2:end))/2;
   
                options = optimoptions('fmincon', 'Display', 'off', 'Algorithm', 'sqp', 'MaxFunctionEvaluations', 10^4);
    
                % 1. минимальное значение дисперсии
                objFun = @(p) (p'*(binMids.^2) - (p'*binMids)^2);
                p_min = fmincon(objFun, ones(nBins,1)/nBins, A, b, Aeq, beq, lb, ub, [], options);
                minVar = objFun(p_min);
                % 2. максимальное значение дисперсии
                objFun = @(p) -(p'*(binMids.^2) + (p'*binMids)^2);
                p_max = fmincon(objFun, ones(nBins,1)/nBins, A, b, Aeq, beq, lb, ub, [], options);
                maxVar = -objFun(p_max);
    
                stdCint = sqrt([minVar, maxVar]);
            else
                error('Аргумент method не задан. Допустимые значения: approximate, accurate.');
            end
        end
        
        % Сложение с использованием принципа Заде (правило max-min)
        function Z = plus(A, B)
            if ~isa(A, 'FuzzyInterval') || ~isa(B, 'FuzzyInterval')
                error('Оба операнда должны быть объектами типа FuzzyInterval.');
            end
            
            % комбинирование и сортировка уровней значимости
            combinedAlpha = unique([A.AlphaLevels, B.AlphaLevels]);
            combinedAlpha = sort(combinedAlpha, 'descend');
            
            % вычисление интервалов на каждом уровне значимости
            sumIntervals = zeros(length(combinedAlpha), 2);
            for i = 1:length(combinedAlpha)
                alpha = combinedAlpha(i);
                intA = A.getIntervalAtAlpha(alpha);
                intB = B.getIntervalAtAlpha(alpha);
                % суммирование интервалов по интервальной арифметике
                sumIntervals(i,:) = [intA(1)+intB(1), intA(2)+intB(2)];
            end
            
            Z = FuzzyInterval(combinedAlpha, sumIntervals);
        end
        
        % вложенный интервал на заданном уровне значимости (с интерполяцией)
        function interval = getIntervalAtAlpha(obj, alpha)
            if alpha < 0 || alpha > 1
                error('Значение alpha должно быть от 0 до 1.');
            end
            
            % граничные значения α
            idxHigh = find(obj.AlphaLevels >= alpha, 1, 'last');
            idxLow = find(obj.AlphaLevels <= alpha, 1, 'first');
            
            if isempty(idxHigh)
                interval = obj.Intervals(1,:);
            elseif isempty(idxLow) 
                interval = obj.Intervals(end,:);
            elseif idxHigh == idxLow  % точное соответствие
                interval = obj.Intervals(idxHigh,:);
            else  % линейная интерполяция
                alphaHigh = obj.AlphaLevels(idxHigh);
                alphaLow = obj.AlphaLevels(idxLow);
                weight = (alpha-alphaLow)/(alphaHigh-alphaLow);
                
                % интерполяция левых границ
                leftHigh = obj.Intervals(idxHigh,1);
                leftLow = obj.Intervals(idxLow,1);
                left = leftHigh + (leftLow-leftHigh)*(1-weight);
                
                % интерполяция правых границ
                rightHigh = obj.Intervals(idxHigh,2);
                rightLow = obj.Intervals(idxLow,2);
                right = rightHigh + (rightLow-rightHigh)*weight;
                
                interval = [left, right];
            end
        end

        % Умножение на коэффициент
        function Z = mtimes(k, A)
            if ~isa(A, 'FuzzyInterval') && ~isa(k, 'double')
                error('Операция умножения задана только для перемножения с коэффициентом.');
            end
            z_alphalevels = A.AlphaLevels;
            z_intervals = k * A.Intervals;
            if k < 0
                z_intervals = z_intervals(:,[2,1]);
            end
            Z = FuzzyInterval(z_alphalevels, z_intervals);
        end

        % Графическое отображение функции принадлежности нечеткого интервала
        function plot(obj)
            figure;
            hold on;
            
            % Отображение α-cuts
            for i = 1:length(obj.AlphaLevels)
                a = obj.Intervals(i,1);
                b = obj.Intervals(i,2);
                alpha = obj.AlphaLevels(i);
                plot([a b], [alpha alpha], 'b-', 'LineWidth', 1.5);
                
                % Отображение вертикальных соединительных линий
                if i < length(obj.AlphaLevels)
                    next_a = obj.Intervals(i+1,1);
                    next_b = obj.Intervals(i+1,2);
                    plot([a next_a], [alpha obj.AlphaLevels(i+1)], 'b--');
                    plot([b next_b], [alpha obj.AlphaLevels(i+1)], 'b--');
                end
            end
            
            xlabel('Значение');
            ylabel('\alpha');
            title('Нечеткий интервал.');
            grid on;
            ylim([0 1+0.1]);
        end

        % преобразование в PBox
        function pbox = FuzzyInterval2PBox(obj, numPoints)
            if nargin < 2
                numPoints = length(obj.AlphaLevels);
            end

            x_vals = linspace(min(obj.Intervals(:)), max(obj.Intervals(:)), numPoints);
            lowerCDF = zeros(size(x_vals));
            upperCDF = zeros(size(x_vals));
            
            for i = 1:length(x_vals)
                x = x_vals(i);
                bound = max(obj.AlphaLevels(obj.Intervals(:,1) < x));
                if ~isempty(bound)
                   upperCDF(i) = min(bound, 1);
                else
                   upperCDF(i) = 0;
                end
                bound = 1 - max(obj.AlphaLevels(obj.Intervals(:,2) > x));
                if ~isempty(bound)
                    lowerCDF(i) = max(bound, 0);
                else 
                    lowerCDF(i) = 1;
                end
            end
            
            pbox = PBox(x_vals, lowerCDF, upperCDF);
        end

        % преобразование в Hist
        function Histogram = FuzzyInterval2Hist(obj, numBins)
            if nargin < 2
                numBins = length(obj.AlphaLevels);
            end
            % определение границ полос гистограммы
            edges = linspace(min(obj.Intervals(:)), max(obj.Intervals(:)), numBins+1);
            
            lowerPDF = zeros(1, numBins);
            upperPDF = zeros(1, numBins);
            
            for i = 1:numBins
                left = edges(i);
                right = edges(i+1);
                
                % определение максимального уровня значимости, при котором интервал содержит данную полосу
                containing = (obj.Intervals(:,1) <= left) & (obj.Intervals(:,2) >= right);
                if any(containing)
                    max_alpha = max(obj.AlphaLevels(containing));
                else
                    max_alpha = 0;
                end
                
                % границы PDF пропорционально уровню значимости
                lowerPDF(i) = 0;  % консервативная нижняя граница
                upperPDF(i) = max_alpha;
            end
            
            % определение границ значений PDF
            bin_widths = diff(edges);
            total_upper = sum(upperPDF .* bin_widths);
            upperPDF = upperPDF / total_upper;
            
            Histogram = Hist(edges, lowerPDF, upperPDF);
        end

        % преобразование в DempsterShafer
        function ds = FuzzyInterval2DempsterShafer(obj, method)
            if (nargin < 2) || strcmp(method, 'fractional')
                % вычисление масс из разностей значений уровня значимости
                masses = -0.5*diff([obj.AlphaLevels']);            
                masses = [masses; masses];
                intervals = [];
                if obj.AlphaLevels(1) == 1
                   masses = [1.0; masses];
                   intervals = [obj.Intervals(1,:); intervals];
                end
                intervals = [intervals; [obj.Intervals(2:end,1), obj.Intervals(1:end-1,1)]];
                intervals = [intervals; [obj.Intervals(1:end-1,2), obj.Intervals(2:end,2)]];
                % удаление интервалов с нулевой массой
                valid = masses > 0;
                ds = DempsterShafer(intervals(valid,:), masses(valid)'/sum(masses));
            elseif strcmp(method, 'nested')
                masses = -diff(obj.AlphaLevels');
                % удаление интервалов с нулевой массой
                valid = masses > 0;
                ds = DempsterShafer(obj.Intervals(valid,:), masses(valid)'/sum(masses(valid)));
            else
                error('Поле method может быть равно только fractional или nested.');
            end
        end

        % преобразование в Fuzzy
        function fv = FuzzyInterval2Fuzzy(obj, numPoints)
            if nargin < 2
                numPoints = length(obj.AlphaLevels);
            end
            % дискретизация носителя
            x_vals = linspace(min(obj.Intervals(:)), max(obj.Intervals(:)), numPoints);
            membership = zeros(size(x_vals));
            
            % оценка уровня значимости как максимального alpha, при котором вложенный интервал содержит x
            for i = 1:length(x_vals)
                x = x_vals(i);
                contains_x = (obj.Intervals(:,1) <= x) & (obj.Intervals(:,2) >= x);
                if any(contains_x)
                    membership(i) = max(obj.AlphaLevels(contains_x));
                end
            end
            
            fv = Fuzzy(x_vals, membership);
        end
    end
end

File: DataReconcile_Python/Uncertainty/FuzzyInterval.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize

# Преобразование разных типов представления неточных данных к формату,
# унифицированному для дальнейшего применения Data Reconciliation.

# Класс для переменных типа FuzzyInterval (нечеткий интервал)
class FuzzyInterval:
    def __init__(self, alphaLevels, intervals):
        # Проверка входных аргументов
        if intervals.shape[1] != 2:
            raise ValueError('Границы вложенных интервалов должны задаваться строкой из двух значений.')
        if len(alphaLevels) != intervals.shape[0]:
            raise ValueError('Количество уровней alpha-cut должно соответствовать числу вложенных интервалов.')
        if np.any((alphaLevels < 0-1e-6) | (alphaLevels > 1+1e-6)):
            raise ValueError('Уровни alpha-cut должны иметь значения в пределах от 0 до 1.')
        if np.any(intervals[:, 0] > intervals[:, 1]):
            raise ValueError('Правые границы вложенных интервалов должны быть меньше левых границ.')

        # Сортировка по значениям alpha-cut
        idx = np.argsort(alphaLevels)[::-1]
        self.AlphaLevels = np.array(alphaLevels)[idx]
        self.Intervals = np.array(intervals)[idx]

        # # Проверка вложенного характера интервалов Intervals:
        # # с уменьшением α-cut вложенные интервалы должны расширяться.
        # if np.any(self.Intervals[:, 0] != np.sort(self.Intervals[:, 0])[::-1]) or \
        #    np.any(self.Intervals[:, 1] != np.sort(self.Intervals[:, 1])):
        #     raise ValueError('Вложенность интервалов в поле Intervals нарушена.')

    # Определение величины возможного среднеквадратического отклонения (согласно теореме Крейновича-Тао).
    def getStd(self, method='approximate'):
        int = self.Intervals  # Переменная для краткости
        a = self.AlphaLevels  # Переменная для краткости
        widths = int[:, 1] - int[:, 0]  # Ширина интервалов

        if method == 'approximate':
            alpha = 0.05
            ind1 = np.where(a == alpha)[0]
            ind2 = np.where(a == 1)[0]
            if ind1.size > 0 and ind2.size > 0:
                stdCint = np.array([0, 0.25 * (widths[ind1[0]] - widths[ind2[0]])])
            else:
                ind1 = np.where(a > alpha)[0]
                if ind1.size == 0:
                    stdCint = np.array([0, 0.25 * widths[0]])
                else:
                    ind1 = ind1[-1]
                    if ind1 == 0:
                        stdCint = np.array([0, 0.25 * widths[ind1]])
                    else:
                        # Линейная интерполяция
                        width = (alpha - a[ind1 - 1]) * (widths[ind1] - widths[ind1 - 1]) / (a[ind1] - a[ind1 - 1]) + widths[ind1 - 1]
                        stdCint = np.array([0, 0.25 * width])
                    if ind2.size > 0:
                        stdCint[1] -= 0.25 * widths[ind2[0]]
        elif method == 'accurate':
            # Дискретизация носителя, используя границы вложенных интервалов
            breakpoints = np.unique(np.concatenate([self.Intervals[:, 0], self.Intervals[:, 1]]))
            nBins = len(breakpoints) - 1
            binEdges = np.sort(breakpoints)

            # Определение ограничений на вероятности из α-cuts
            A = []
            b = []
            for i in range(len(self.AlphaLevels)):
                alpha = self.AlphaLevels[i]
                a_int = self.Intervals[i, 0]
                b_int = self.Intervals[i, 1]
                inBin = (binEdges[:-1] >= a_int) & (binEdges[1:] <= b_int)

                # Добавление ограничения: sum(p(inBin)) >= 1 - alpha
                A.append(-inBin.astype(float))
                b.append(-(1 - alpha))

            A = np.array(A)
            b = np.array(b)

            # Добавление ограничений на сумму вероятностей
            Aeq = np.ones((1, nBins))
            beq = np.array([1])
            # Добавление ограничения на неотрицательность вероятностей
            lb = np.zeros(nBins)
            ub = None

            # Параметры промежутков
            binMids = (binEdges[:-1] + binEdges[1:]) / 2

            # 1. Минимальное значение дисперсии
            def objFun(p):
                return np.dot(p, binMids**2) - np.dot(p, binMids)**2

            p0 = np.ones(nBins) / nBins
            res = minimize(objFun, p0, constraints={'type': 'ineq', 'fun': lambda p: b - np.dot(A, p)},
                          bounds=[(0, None)] * nBins, options={'maxiter': 10000, 'disp': False})
            minVar = objFun(res.x)

            # 2. Максимальное значение дисперсии
            def objFunMax(p):
                return -(np.dot(p, binMids**2) - np.dot(p, binMids)**2)

            res = minimize(objFunMax, p0, constraints={'type': 'ineq', 'fun': lambda p: b - np.dot(A, p)},
                          bounds=[(0, None)] * nBins, options={'maxiter': 10000, 'disp': False})
            maxVar = -objFunMax(res.x)

            stdCint = np.sqrt([minVar, maxVar])
        else:
            raise ValueError('Аргумент method не задан. Допустимые значения: approximate, accurate.')

        return stdCint

    # Сложение с использованием принципа Заде (правило max-min)
    def __add__(self, other):
        if not isinstance(other, FuzzyInterval):
            raise TypeError('Оба операнда должны быть объектами типа FuzzyInterval.')

        # Комбинирование и сортировка уровней значимости
        combinedAlpha = np.unique(np.concatenate([self.AlphaLevels, other.AlphaLevels]))
        combinedAlpha = np.sort(combinedAlpha)[::-1]

        # Вычисление интервалов на каждом уровне значимости
        sumIntervals = np.zeros((len(combinedAlpha), 2))
        for i, alpha in enumerate(combinedAlpha):
            intA = self.getIntervalAtAlpha(alpha)
            intB = other.getIntervalAtAlpha(alpha)
            # Суммирование интервалов по интервальной арифметике
            sumIntervals[i] = [intA[0] + intB[0], intA[1] + intB[1]]

        return FuzzyInterval(combinedAlpha, sumIntervals)

    # Вложенный интервал на заданном уровне значимости (с интерполяцией)
    def getIntervalAtAlpha(self, alpha):
        if alpha < 0 or alpha > 1:
            raise ValueError('Значение alpha должно быть от 0 до 1.')

        # Граничные значения α
        idxHigh = np.where(self.AlphaLevels >= alpha)[0]
        idxLow = np.where(self.AlphaLevels <= alpha)[0]

        if idxHigh.size == 0:
            return self.Intervals[0]
        elif idxLow.size == 0:
            return self.Intervals[-1]
        elif idxHigh[-1] == idxLow[0]:  # Точное соответствие
            return self.Intervals[idxHigh[-1]]
        else:  # Линейная интерполяция
            idxHigh = idxHigh[-1]
            idxLow = idxLow[0]
            alphaHigh = self.AlphaLevels[idxHigh]
            alphaLow = self.AlphaLevels[idxLow]
            weight = (alpha - alphaLow) / (alphaHigh - alphaLow)

            # Интерполяция левых границ
            leftHigh = self.Intervals[idxHigh, 0]
            leftLow = self.Intervals[idxLow, 0]
            left = leftHigh + (leftLow - leftHigh) * (1 - weight)

            # Интерполяция правых границ
            rightHigh = self.Intervals[idxHigh, 1]
            rightLow = self.Intervals[idxLow, 1]
            right = rightHigh + (rightLow - rightHigh) * weight

            return np.array([left, right])

    # Умножение на коэффициент
    def __rmul__(self, k):
        if not isinstance(k, (int, float)):
            raise TypeError('Операция умножения задана только для перемножения с коэффициентом.')
        z_alphalevels = self.AlphaLevels
        z_intervals = k * self.Intervals
        if k < 0:
            z_intervals = z_intervals[:, [1, 0]]
        return FuzzyInterval(z_alphalevels, z_intervals)

    # Графическое отображение функции принадлежности нечеткого интервала
    def plot(self):
        plt.figure()

        # Отображение α-cuts
        for i in range(len(self.AlphaLevels)):
            a = self.Intervals[i, 0]
            b = self.Intervals[i, 1]
            alpha = self.AlphaLevels[i]
            plt.plot([a, b], [alpha, alpha], 'b-', linewidth=1.5)

            # Отображение вертикальных соединительных линий
            if i < len(self.AlphaLevels) - 1:
                next_a = self.Intervals[i + 1, 0]
                next_b = self.Intervals[i + 1, 1]
                plt.plot([a, next_a], [alpha, self.AlphaLevels[i + 1]], 'b--')
                plt.plot([b, next_b], [alpha, self.AlphaLevels[i + 1]], 'b--')

        plt.xlabel('Значение')
        plt.ylabel('α')
        plt.title('Нечеткий интервал.')
        plt.grid(True)
        plt.ylim([0, 1.1])
        plt.show()

    # Преобразование в PBox
    def FuzzyInterval2PBox(self, numPoints=None):
        if numPoints is None:
            numPoints = len(self.AlphaLevels)
        x_vals = np.linspace(np.min(self.Intervals), np.max(self.Intervals), numPoints)
        lowerCDF = np.zeros_like(x_vals)
        upperCDF = np.zeros_like(x_vals)

        for i, x in enumerate(x_vals):
            bound = np.max(self.AlphaLevels[self.Intervals[:, 0] < x], initial=0)
            upperCDF[i] = min(bound, 1) if bound.size > 0 else 0
            bound = 1 - np.max(self.AlphaLevels[self.Intervals[:, 1] > x], initial=0)
            lowerCDF[i] = max(bound, 0) if bound.size > 0 else 1

        return PBox(x_vals, lowerCDF, upperCDF)

    # Преобразование в Hist
    def FuzzyInterval2Hist(self, numBins=None):
        if numBins is None:
            numBins = len(self.AlphaLevels)
        # Определение границ полос гистограммы
        edges = np.linspace(np.min(self.Intervals), np.max(self.Intervals), numBins + 1)

        lowerPDF = np.zeros(numBins)
        upperPDF = np.zeros(numBins)

        for i in range(numBins):
            left = edges[i]
            right = edges[i + 1]

            # Определение максимального уровня значимости, при котором интервал содержит данную полосу
            containing = (self.Intervals[:, 0] <= left) & (self.Intervals[:, 1] >= right)
            if np.any(containing):
                max_alpha = np.max(self.AlphaLevels[containing])
            else:
                max_alpha = 0

            # Границы PDF пропорционально уровню значимости
            lowerPDF[i] = 0  # Консервативная нижняя граница
            upperPDF[i] = max_alpha

        # Определение границ значений PDF
        bin_widths = np.diff(edges)
        total_upper = np.sum(upperPDF * bin_widths)
        upperPDF = upperPDF / total_upper

        return Hist(edges, lowerPDF, upperPDF)

    # Преобразование в DempsterShafer
    def FuzzyInterval2DempsterShafer(self, method='fractional'):
        if method == 'fractional':
            # Вычисление масс из разностей значений уровня значимости
            masses = -0.5 * np.diff(np.concatenate([[0], self.AlphaLevels]))
            masses = np.concatenate([masses, masses])
            intervals = []
            if self.AlphaLevels[0] == 1:
                masses = np.concatenate([[1.0], masses])
                intervals = [self.Intervals[0]]
            intervals = np.concatenate([intervals, self.Intervals[1:, 0].reshape(-1, 1), self.Intervals[:-1, 0].reshape(-1, 1)], axis=1)
            intervals = np.concatenate([intervals, self.Intervals[:-1, 1].reshape(-1, 1), self.Intervals[1:, 1].reshape(-1, 1)], axis=1)
            # Удаление интервалов с нулевой массой
            valid = masses > 0
            ds = DempsterShafer(intervals[valid], masses[valid] / np.sum(masses[valid]))
        elif method == 'nested':
            masses = -np.diff(np.concatenate([[0], self.AlphaLevels]))
            # Удаление интервалов с нулевой массой
            valid = masses > 0
            ds = DempsterShafer(self.Intervals[valid], masses[valid] / np.sum(masses[valid]))
        else:
            raise ValueError('Поле method может быть равно только fractional или nested.')

        return ds

    # Преобразование в Fuzzy
    def FuzzyInterval2Fuzzy(self, numPoints=None):
        if numPoints is None:
            numPoints = len(self.AlphaLevels)
        # Дискретизация носителя
        x_vals = np.linspace(np.min(self.Intervals), np.max(self.Intervals), numPoints)
        membership = np.zeros_like(x_vals)

        # Оценка уровня значимости как максимального alpha, при котором вложенный интервал содержит x
        for i, x in enumerate(x_vals):
            contains_x = (self.Intervals[:, 0] <= x) & (self.Intervals[:, 1] >= x)
            if np.any(contains_x):
                membership[i] = np.max(self.AlphaLevels[contains_x])

        return Fuzzy(x_vals, membership)
